pinot-core/src/main/java/org/apache/pinot/core/data/table/OrderedIndexedTableResizer.java

```diff
@@ -0,0 +1,260 @@
+      List<SelectionSort> orderBy) {
+
+    int numAggregations = aggregationInfos.size();
+    int numKeyColumns = dataSchema.size() - numAggregations;
+
+    Map<String, Integer> keyIndexMap = new HashMap<>();
+    for (int i = 0; i < numKeyColumns; i++) {
+      keyIndexMap.put(dataSchema.getColumnName(i), i);
+    }
+
+    Map<String, Integer> aggregationColumnToIndex = new HashMap<>();
+    Map<String, AggregationInfo> aggregationColumnToInfo = new HashMap<>(aggregationInfos.size());
+    for (int i = 0; i < numAggregations; i++) {
+      AggregationInfo aggregationInfo = aggregationInfos.get(i);
+      String aggregationColumn = AggregationFunctionUtils.getAggregationColumnName(aggregationInfo);
+      aggregationColumnToIndex.put(aggregationColumn, i);
+      aggregationColumnToInfo.put(aggregationColumn, aggregationInfo);
+    }
+
+    _numOrderBy = orderBy.size();
+    _orderByInfos = new ArrayList<>(_numOrderBy);
+    Comparator[] comparators = new Comparator[_numOrderBy];
+
+    for (int i = 0; i < _numOrderBy; i++) {
+      SelectionSort selectionSort = orderBy.get(i);
+      String column = selectionSort.getColumn();
+
+      if (keyIndexMap.containsKey(column)) {
+        int index = keyIndexMap.get(column);
+        _orderByInfos.add(createOrderByInfoForKeyColumn(index));
+      } else if (aggregationColumnToIndex.containsKey(column)) {
+        int index = aggregationColumnToIndex.get(column);
+        _orderByInfos.add(createOrderByInfoForAggregationColumn(index, aggregationColumnToInfo.get(column)));
+      } else {
+        throw new UnsupportedOperationException("Could not find column " + column + " in data schema");
+      }
+
+      comparators[i] = ComparableComparator.getInstance();
+      if (!selectionSort.isIsAsc()) {
+        comparators[i] = comparators[i].reversed();
+      }
+    }
+
+    _intermediateRecordComparator = (o1, o2) -> {
+
+      for (int i = 0; i < _numOrderBy; i++) {
+        int result = comparators[i].compare(o1.getValues()[i], o2.getValues()[i]);
+        if (result != 0) {
+          return result;
+        }
+      }
+      return 0;
+    };
+  }
+
+  /**
+   * Constructs an IntermediateRecord from Record
+   * The IntermediateRecord::key is the same Record::key
+   * The IntermediateRecord::values contains only the order by columns, in the query's sort sequence
+   * For aggregation values in the order by, the final result is extracted if the intermediate result is non-comparable
+   */
+  @VisibleForTesting
+  IntermediateRecord getIntermediateRecord(Record record) {
+    Comparable[] intermediateRecordValues = new Comparable[_numOrderBy];
+
+    Object[] keyColumns = record.getKey().getColumns();
+    Object[] aggregationColumns = record.getValues();
+    for (int i = 0; i < _numOrderBy; i++) {
+      OrderByInfo orderByInfo = _orderByInfos.get(i);
+      Comparable comparable;
+      if (orderByInfo.getOrderByType().equals(OrderByType.KEY_COLUMN)) {
```

Comment:
To avoid these if checks, we should add an inner class to extract comparable from the Record. E.g. have a KeyExtractor and an AggregationResultExtractor

Reply:
Great suggestion. Added an inner class `OrderByValueExtractor` with implementations `KeyColumnExtractor` and `ComparableAggregationColumnExtractor` and `NonComparableAggregationColumnExtractor`. Separated these 2 to further avoid the if check `if (aggregationFunction.isIntermediateResultComparable()`